### ISSUE TITLE
This warning about a trivial numeric cast feels wrong

### ISSUE THREAD

**Issue by tommit** [rust-lang/rust#23720]
Intuitively it feels wrong for the compiler to warn about a trivial numeric cast in a context where the act of removing the cast would result in an error.

```rust
macro_rules! make_foo {
    ($Int:ident) => (
        #[repr($Int)]
        enum Foo { A, B, C }

        impl Foo {
            fn min() -> Foo {
                unsafe {
                    ::std::mem::transmute(0 as $Int) // WARNING: [1]
                }
            }
        }
    )
}

make_foo!{ u8 }

// [1] warning: trivial numeric cast: `u8` as `u8`, #[warn(trivial_numeric_casts)] on by default
```

After removing the cast:

```rust
macro_rules! make_foo {
    ($Int:ident) => (
        #[repr($Int)]
        enum Foo { A, B, C }

        impl Foo {
            fn min() -> Foo {
                unsafe {
                    ::std::mem::transmute(0) // ERROR: [2]
                }
            }
        }
    )
}

make_foo!{ u8 }

// [2] error: transmute called on types with different sizes: i32 (32 bits) to Foo (8 bits)
```

**Comment by eddyb** [comment]
You should provide type parameters to `transmute` wherever inference is not enough - it's less dangerous that way, too:
```rust
macro_rules! make_foo {
    ($Int:ident) => (
        #[repr($Int)]
        enum Foo { A, B, C }

        impl Foo {
            fn min() -> Foo {
                unsafe {
                    ::std::mem::transmute::<$Int, Foo>(0)
                }
            }
        }
    )
}

make_foo!{ u8 }
```
